### PR TITLE
eks-prow-build-cluster: set permission boundary for roles created via EKS and IRSA modules

### PR DESCRIPTION
**Pull request by pkprzekwas**
Add permission boundaries to roles previously created with EKS and IRSA terraform modules. Definition of permission boundary is located [here](https://github.com/kubernetes/k8s.io/blob/main/infra/aws/terraform/modules/eks-prow-iam/boundary.tf).

/assign @xmudrii 
/cc @sftim

### CODE REVIEW

**Comment by pkprzekwas** [comment]
/hold
wait for applying on k8s-infra-prow account

**Review by xmudrii — APPROVED** [review]
/lgtm
/approve
/hold to apply to prod

File: infra/aws/terraform/modules/eks-prow-iam/policies.tf [hunk 1]
```diff
@@ -143,6 +143,8 @@ data "aws_iam_policy_document" "eks_apply" {
       "iam:TagPolicy",
       "iam:TagRole",
       "iam:UpdateOpenIDConnectProviderThumbprint",
+      "iam:PutRolePermissionsBoundary",
+      "iam:DeleteRolePermissionsBoundary",
```
Comment (xmudrii) [inline infra/aws/terraform/modules/eks-prow-iam/policies.tf:147]: Should `DeleteRole*` go here or in the destroy policy?

File: infra/aws/terraform/prow-build-cluster/eks.tf [hunk 2]
```diff
@@ -44,6 +44,8 @@ module "eks" {
     },
   ]
 
+  iam_role_permissions_boundary = "arn:aws:iam::${local.account_id}:policy/ProvisionerPermissionBoundary"
```
Comment (xmudrii) [inline infra/aws/terraform/prow-build-cluster/eks.tf:47]: Is it possible to automatically figure out this ARN with data resource? Doing that will also ensure that this permission boundary exists.

**Review by xmudrii — COMMENTED** [review]
We should add [`permission_boundary`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_role#permissions_boundary) to all IAM roles that we create via Terraform. It's going to be mandatory once we start enforcing our management IAM role to create IAM roles with the permission boundary.

**Comment by sftim** [comment]
We might want a break-glass role that's allowed to edit the boundary policy, and guidelines around using it.

File: infra/aws/terraform/modules/eks-prow-iam/policies.tf (hunk 1 above)
Comment (pkprzekwas) [reply infra/aws/terraform/modules/eks-prow-iam/policies.tf:147]: Good point

File: infra/aws/terraform/prow-build-cluster/eks.tf (hunk 2 above)
Comment (pkprzekwas) [reply infra/aws/terraform/prow-build-cluster/eks.tf:47]: Yeah, that's a great idea

**Comment by pkprzekwas** [comment]
> We might want a break-glass role that's allowed to edit the boundary policy, and guidelines around using it.

E.g. `InfraAdmin` role that could be assumed just by very narrow set of users and would contain `AdministratorAccess` policy?

**Comment by sftim** [comment]
> E.g. `InfraAdmin` role that could be assumed just by very narrow set of users and would contain `AdministratorAccess` policy?

That's quite broad!

The role I'm imagining could have an inline policy that _only_ allows:
- IAM read access
- the ability to modify the IAM policies that we use for permissions boundaries

It depends what we want to achieve. Allowing `*:*` really is a lot of access.

**Comment by xmudrii** [comment]
I'm wondering if we should have such role at all. Instead, I think SIG leads should have access to the root account (shared via 1Password). They can then create an IAM user with needed access (e.g. AdministratorAccess). That way, we reduce potential attack surface (you can't assume role if there's no such role). This is based on assumption that break-glass access should be needed very rare (if at all).

**Comment by pkprzekwas** [comment]
I think access to the root account may not be sufficient in this case. Or maybe you meant root access to the k8s-infra-prow account?

**Comment by xmudrii** [comment]
@pkprzekwas Yes, I meant root access to the k8s-infra-prow account.

**Review by xmudrii — APPROVED** [review]
/lgtm
/approve

**Comment by pkprzekwas** [comment]
All changes applied on prow-build-cluster.

**Review by xmudrii — APPROVED** [review]

**Pull request merged**